Library/Homebrew/extend/os/mac/formula_cellar_checks.rb

```diff
@@ -56,10 +58,42 @@ def check_python_framework_links(lib)
     EOS
   end
 
+  def check_linkage
+    return unless formula.prefix.directory?
+    keg = Keg.new(formula.prefix)
+    checker = LinkageChecker.new(keg, formula)
+    s = ""
+
+    if checker.broken_dylibs?
+      s += <<-EOS.undent
+        Broken dylib links found:
+          #{checker.broken_dylibs.to_a * "\n          "}
+
+        It is likely the installation was broken.
+        Please check formula and make sure it is built from source.
+      EOS
+    end
+
+    if checker.undeclared_deps?
+      s += <<-EOS.undent
+        Undeclared dependencies found:
+          #{checker.undeclared_deps * "\n          "}
+
+        Formulae are required to declare all linked dependencies.
+        Please check whether these dependencies are indeed needed.
+        If so, please add all missing dependencies to the formula with:
```

Comment:
>we can reevaluate it if it encourages bad behaviour.

It's not bad behavior (unless you mean the software not the people). The contributors who are "following the rules" will run brew audit locally, and it will expressly tell them to add garbage. That exact same command run during CI would not have done so.